examples/chainermn/cifar/train_cifar.py

```diff
@@ -0,0 +1,138 @@
+import argparse
+
+import chainermn
+import chainer
+import chainer.links as L
+from chainer import training
+from chainer.training import extensions
+from chainer.training import triggers
+
+from chainer.datasets import get_cifar10
+from chainer.datasets import get_cifar100
+
+import models.VGG
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Chainer CIFAR example:')
+    parser.add_argument('--dataset', '-d', default='cifar10',
+                        help='The dataset to use: cifar10 or cifar100')
+    parser.add_argument('--batchsize', '-b', type=int, default=64,
+                        help='Number of images in each mini-batch')
+    parser.add_argument('--learnrate', '-l', type=float, default=0.05,
+                        help='Learning rate for SGD')
+    parser.add_argument('--epoch', '-e', type=int, default=300,
+                        help='Number of sweeps over the dataset to train')
+    parser.add_argument('--gpu', '-g', action='store_true',  
+                        help='GPU ID (negative value indicates CPU)')
+    parser.add_argument('--out', '-o', default='result',
+                        help='Directory to output the result')
+    parser.add_argument('--resume', '-r', default='',
+                        help='Resume the training from snapshot')
+    parser.add_argument('--early-stopping', type=str,
+                        help='Metric to watch for early stopping')
+    args = parser.parse_args()
+
+    # Prepare ChainerMN communicator.
+    if args.gpu:
+        comm = chainermn.create_communicator('hierarchical')
+        data_axis, model_axis = comm.rank % 2, comm.rank // 2
+        data_comm = comm.split(data_axis, comm.rank)
+        model_comm = comm.split(model_axis, comm.rank)
+        device = comm.intra_rank
+    else:
+        comm = chainermn.create_communicator('naive')
+        data_axis, model_axis = comm.rank % 2, comm.rank // 2
```

Comment:
I don't quite understand what `data_comm` and `model_comm` are for.

Why the dataset is loaded using `model_comm`, and `scatter_dataset` is used over `data_comm` ? Why don't you just let rank 0 load the dataset, and simply `scatter` it?
I believe that there must be 4 processes, because `model_comm.size == 2`, and the restriction does not make much sense to me.

Then, `data_axis` and `model_axis` are:

| rank | data_axis | model_axis |
|:---|:---|:---|
|0 | 0 | 0 |
|1 | 1 | 0 |
|2 | 0 | 1 |
|3 | 1 | 1 |

As far as I understand, what happens is (all ranks below are all in `COMM_WORLD`)

 * rank 0 and 1 load the dataset (because `model_axis` are `0`)
 * rank 0 (1) splits the data and sends the half to rank 2 (3) (because 0(1) and 2(3) have `data_axis == 0 (1)` thus belong to the same `data_comm`).

The issues are:

+ The behavior is very confusing and hard to understand.
+ There are two copies of CIFAR datasets on the processes, and each copy is split by two processes. Although the training works, it is inefficient and confusing.
 + The number of process is limited to 4, which I guess is unnecessary restriction.

I would suggest to just simply use `COMM_WORLD`. 

NOTE:
This is just my guessing: `scatter_dataset` does not need to be hierarchical, even if the communicator is `hierarchical`. `hierarchical` implies the *internal* implementation inside a communicator is hierarchical, so it does not affect the use of communicators.

Reply:
Fix simply use `COMM_WORLD`.
And add argument `--communicator`.